### PR TITLE
Fixing the bug regurding tuple types.

### PR DESCRIPTION
**Pull request by FlaminGuy**
This will iterate over types of functions too.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
1. Does this problem exist for things other than tuples? E.g. lambdas, enums, etc?
2. Using `:::` is going to be pretty slow. Can you use a set instead or is the order important?

**Comment by FlaminGuy** [comment]
1. I already get all enums types lifted in the `root` object passed into to the phase so this isn't a problem for enums
2. We can do set

**Comment by FlaminGuy** [comment]
I also added a very quick commit for adding line numbers and locations. Those are just for debugging and when an error happens, otherwise they have no effect on anything.

File: main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala [hunk 1]
```diff
@@ -244,6 +244,9 @@ object EnumGen extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
     val implementedInterfaces = Array(decorate(superType))
     visitor.visit(JavaVersion, ACC_PUBLIC + ACC_FINAL, decorate(className), null, superClass, implementedInterfaces)
 
+    // Source of the class
+    visitor.visitSource(decorate(className), null)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala]: As above, I would like us to use the filename.

File: main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala [hunk 2]
```diff
@@ -309,24 +312,32 @@ object CodeGen extends Phase[ExecutableAst.Root, ExecutableAst.Root]{
...
+             Type.Apply(Type.Enum(_, _), _) => visitor.visitVarInsn(ALOAD, sym.getStackOffset)
+        case _ if tpe.isTuple => visitor.visitVarInsn(ALOAD, sym.getStackOffset)
+        case _ => throw InternalCompilerException(s"Unexpected type: `$tpe'.")
+      }
     }
 
-    case Expression.Ref(name, _, _) =>
+    case Expression.Ref(name, _, loc) =>
+      // Adding source line number for debugging
+      addSourceLine(visitor,loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: Nice! Good use of a little nifty helper :)

File: main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala [hunk 3]
```diff
@@ -184,6 +184,9 @@ object CodeGen extends Phase[ExecutableAst.Root, ExecutableAst.Root]{
     visitor.visit(JavaVersion, ACC_PUBLIC + ACC_SUPER, decorate(prefix), null,
       asm.Type.getInternalName(Constants.objectClass), null)
 
+    // Source of the class
+    visitor.visitSource(decorate(prefix), null)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: Just the base filename, not the full path.

**Comment by magnus-madsen** [comment]
I am sorry for breaking your PR :|

The changes look good to me. Elegant the way you added source information, I like it :)

I wonder if we can use the intellij debugger with Flix? If you manually fill in the breakpoint box.

**Pull request merged**